File: Pregunta1/E1.hs
--Ejercicio 1
--haskell
-- Fibonnaci base 3 Estructural
suma3 :: Integer -> Integer -> Integer -> Integer
suma3 a b c = a + b + c 

fiboEst :: Integer -> Integer -> Integer -> Integer -> IO() 
fiboEst n a b c = do
    if n >= 0
    then do
       let f = suma3 a b c
       putStr $ (show(f)++" ")
       fiboEst (n-1) b c f
    else putStr ""

-- Fibonnaci base 3 recursivo
fiboRec :: Integer -> Integer
fiboRec 0 = 0
fiboRec 1 = 1
fiboRec 2 = 1
fiboRec n = fiboRec(n-1) + fiboRec(n-2) + fiboRec(n-3)

generarFiboRec :: Integer -> IO() 
generarFiboRec n = do
    if n >= 0
    then do
       generarFiboRec (n-1)
       putStr $ (show(fiboRec n)++" ")
    else putStr ""

main :: IO()
main = do
    putStrLn ("Ingrese numero de terminos: ")
    n <- getLine
    putStrLn ("Fibonnaci base 3 forma estructural")
    fiboEst (read n) (-1) 1 0
    putStrLn ("\nFibonnaci base 3 forma recursivo")
    generarFiboRec (read n)

File: Pregunta1/E1.scala
/*
Ejericicio 1
scala
*/
object Main {
  // Fibonnaci base 3 forma estructurada
  def fiboEst(n:Int): Int = {
    var f = 0
    var a = -1
    var b = 1
    var c = 0
    for(w <- 0 to n){
      f = a + b + c
      a = b
      b = c
      c = f
      print(f)
      print(" ")
    }
    return 0
  }

  // Fibonnaci base 3 forma recursiva
  def fiboRec(n:Int): Int = {
      if(n >= 0) {
        fiboRec(n-1)
        print(fibo(n))
        print(" ")
      }
      return 0
  }

  def fibo(n:Int): Int = {
    if (n == 0) 
        return 0
    else if( n == 1 || n == 2)
        return 1
    else
        return fibo(n-1) + fibo(n-2) + fibo(n-3)
  }


  def main(args:Array[String]):Unit = {
      println("Generar los n terminos:")
      var n = scala.io.StdIn.readInt()
      println("Fibonnaci base 3 forma estructurada")
      fiboEst(n)
      println("\nFibonnaci base 3 forma recursiva")
      fiboRec(n)
  }
}
